### PR TITLE
Add freeze payment guards

### PR DESCRIPTION
**Pull request by febo**
This PR add a new payment guard with a freeze functionality.

### CODE REVIEW

File: program/src/guards/freeze_sol_payment.rs [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+
+        let seeds = [
+            b"freeze_sol_payment".as_ref(),
+            self.destination.as_ref(),
+            candy_guard_key.as_ref(),
+            candy_machine_key.as_ref(),
+        ];
+
+        let (_, bump) = Pubkey::find_program_address(&seeds, &crate::ID);
+        let signer = [seeds[0], seeds[1], seeds[2], seeds[3], &[bump]];
```
Comment (stegaBOB) [inline program/src/guards/freeze_sol_payment.rs:395]: Is there a more idiomatic way to do this?

File: program/src/guards/freeze_sol_payment.rs [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+            };
+
+        let freeze_guard = if let Some(freeze_guard) = &guard_set.freeze_sol_payment {
+            freeze_guard
+        } else {
+            return err!(CandyGuardError::FreezeGuardNotEnabled);
+        };
+
+        match instruction {
+            // Initializes the freese escrow PDA.
```
Comment (stegaBOB) [inline program/src/guards/freeze_sol_payment.rs:69]:
```suggestion
            // Initializes the freeze escrow PDA.
```

File: program/src/guards/freeze_sol_payment.rs [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+    }
+
+    fn mask() -> u64 {
+        0b1u64 << 16
+    }
+
+    /// Instructions to interact with the freeze feature:
+    ///
+    ///  * initialize
+    ///  * thawn
```
Comment (stegaBOB) [inline program/src/guards/freeze_sol_payment.rs:47]:
```suggestion
    ///  * thaw
```

File: program/src/guards/freeze_sol_payment.rs [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+            //
+            //   0. `[writable]` Freeze PDA to receive the funds (seeds `["freeze_sol_payment",
+            //                   destination pubkey, candy guard pubkey, candy machine pubkey]`).
+            //   1. `[]` Mint account for the NFT.
+            //   2. `[]` Address of the owner of the NFT.
+            //   3. `[writable]` Associate token account of the NFT.
+            //   4. `[]` Master Edition account of the NFT.
+            //   5. `[]` spl-token program ID.
+            //   6. `[]` Metaplex Token Metadata program ID.
+            FreezeInstruction::Thaw => {
```
Comment (stegaBOB) [inline program/src/guards/freeze_sol_payment.rs:162]: No way to thaw if guard is closed. There should be a handler that detects if the assigned guard has been closed and if so, allows thawing and unlocking.

File: program/src/guards/freeze_sol_payment.rs [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+            //
+            // List of accounts required:
+            //
+            //   0. `[writable]` Freeze PDA to receive the funds (seeds `["freeze_sol_payment",
+            //                   destination pubkey, candy guard pubkey, candy machine pubkey]`).
+            //   1. `[signer]` Candy Guard authority.
+            //   2. `[writable]` Address to receive the funds (must match the `destination` address
+            //                   of the guard configuration).
+            //   3. `[]` System program account.
+            FreezeInstruction::UnlockFunds => {
```
Comment (stegaBOB) [inline program/src/guards/freeze_sol_payment.rs:253]: No way to use if guard is closed

File: program/src/guards/freeze_sol_payment.rs [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+            //   1. `[signer]` Candy Guard authority.
+            //   2. `[]` System program account.
+            FreezeInstruction::Initialize => {
+                msg!("FreezeSolPayment: initialize");
+
+                let candy_guard_key = &ctx.accounts.candy_guard.key();
+                let candy_machine_key = &ctx.accounts.candy_machine.key();
+
+                let seeds = [
+                    b"freeze_sol_payment".as_ref(),
```
Comment (stegaBOB) [inline program/src/guards/freeze_sol_payment.rs:84]: `freeze_sol_payment` should probably be a constant. I'd impl a const called `PREFIX` on `FreezeEscrow`. It looks like a lot of other PDAs might not use constants in guard either, so perhaps this should be a separate issue to potentially update all the PDAs to reference constants as the seed prefix.

**Review by stegaBOB — CHANGES_REQUESTED** [review]
Theres a couple potential issues to discuss.

File: program/src/state/candy_guard.rs [hunk 7]
```diff
@@ -99,6 +99,8 @@ pub struct GuardSet {
     pub nft_burn: Option<NftBurn>,
     /// Token burn guard (burn a specified amount of spl-token).
     pub token_burn: Option<TokenBurn>,
+    /// Freeze sol payment (set the price for the mint in lamports with a freeze period).
+    pub freeze_sol_payment: Option<FreezeSolPayment>,
```
[the remark this answers is not in the document]
Comment (febo) [reply program/src/state/candy_guard.rs:103]: Done.

File: program/src/guards/freeze_sol_payment.rs [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+    instruction::{approve, revoke},
+    state::Account as TokenAccount,
+};
+
+use crate::{
+    errors::CandyGuardError,
+    utils::{assert_is_ata, assert_keys_equal, cmp_pubkeys},
+};
+
+/// Guard that charges an amount in SOL (lamports) for the mint.
```
[the remark this answers is not in the document]
Comment (febo) [reply program/src/guards/freeze_sol_payment.rs:21]: Done.

File: program/src/guards/freeze_sol_payment.rs (hunk 2 above)
Comment (febo) [reply program/src/guards/freeze_sol_payment.rs:69]: Done.

File: program/src/guards/freeze_sol_payment.rs (hunk 3 above)
Comment (febo) [reply program/src/guards/freeze_sol_payment.rs:47]: Done.

File: program/src/guards/freeze_sol_payment.rs (hunk 6 above)
Comment (febo) [reply program/src/guards/freeze_sol_payment.rs:84]: Done.

File: program/src/guards/freeze_sol_payment.rs (hunk 1 above)
Comment (febo) [reply program/src/guards/freeze_sol_payment.rs:395]: Changed to:
```
let signer = [
            FreezeEscrow::PREFIX_SEED,
            self.destination.as_ref(),
            candy_guard_key.as_ref(),
            candy_machine_key.as_ref(),
            &[bump],
        ];
```

File: program/src/guards/freeze_sol_payment.rs (hunk 4 above)
Comment (febo) [reply program/src/guards/freeze_sol_payment.rs:102]: Changed `route` instruction implementation so now it is possible to thaw with a closed account.

File: program/src/guards/freeze_token_payment.rs [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+            }
+            FreezeInstruction::Thaw => {
+                msg!("Instruction: Thaw (FreezeTokenPayment guard)");
+                thaw_nft(ctx, route_context, data)
+            }
+
+            FreezeInstruction::UnlockFunds => {
+                msg!("Instruction: Unlock Funds (FreezeTokenPayment guard)");
+                unlock_funds(ctx, route_context)
+            }
```
Comment (stegaBOB) [inline program/src/guards/freeze_token_payment.rs:138]: Missing the list of required accounts comments. I think it should be listed here even though its already stated above the `unlock_funds` method and in teh `thaw_nft` ix in the sol payment

File: program/src/guards/freeze_token_payment.rs [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+    ];
+    let (pda, bump) = Pubkey::find_program_address(&seeds, &crate::ID);
+    assert_keys_equal(freeze_pda.key, &pda)?;
+
+    // authority must the a signer
+    let authority = get_account_info(ctx, 1)?;
+
+    // if the candy guard account is present, we check the authority against
+    // the candy guard authority; otherwise we use the freeze escrow authority
+    let authority_ckeck = if let Some(candy_guard) = route_context.candy_guard {
```
Comment (stegaBOB) [inline program/src/guards/freeze_token_payment.rs:275]: ckeck -> check

File: program/src/guards/freeze_sol_payment.rs [hunk 11]
```diff
@@ -0,0 +1,613 @@
...
+    ];
+    let (pda, _) = Pubkey::find_program_address(&seeds, &crate::ID);
+    assert_keys_equal(freeze_pda.key, &pda)?;
+
+    // authority must the a signer
+    let authority = get_account_info(ctx, 1)?;
+
+    // if the candy guard account is present, we check the authority against
+    // the candy guard authority; otherwise we use the freeze escrow authority
+    let authority_ckeck = if let Some(candy_guard) = route_context.candy_guard {
```
Comment (stegaBOB) [inline program/src/guards/freeze_sol_payment.rs:591]: ckeck -> check

File: program/src/guards/allow_list.rs [hunk 12]
```diff
@@ -74,6 +74,12 @@ impl Guard for AllowList {
         let user = ctx.accounts.payer.key();
         let leaf = solana_program::keccak::hashv(&[user.to_string().as_bytes()]);
 
+        let guard_set = if let Some(guard_set) = route_context.guard_set {
+            guard_set
+        } else {
+            return err!(CandyGuardError::AllowedListNotEnabled);
+        };
```
[the remark this answers is not in the document]
Comment (febo) [reply program/src/guards/allow_list.rs:85]: Good call – added a test for it.

File: program/src/guards/freeze_token_payment.rs (hunk 9 above)
Comment (febo) [reply program/src/guards/freeze_token_payment.rs]: Added comments.

File: program/src/guards/freeze_sol_payment.rs (hunk 11 above)
Comment (febo) [reply program/src/guards/freeze_sol_payment.rs]: Done.

File: program/src/guards/freeze_token_payment.rs (hunk 10 above)
Comment (febo) [reply program/src/guards/freeze_token_payment.rs]: Done.

**Review by stegaBOB — DISMISSED** [review]

**Pull request merged**